adapters/connectad/connectad.go

```diff
@@ -0,0 +1,213 @@
+	headers.Add("Accept", "application/json")
+	addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+	addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+	addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+	if request.Device.DNT != nil {
+		addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
+	} else {
+		addHeaderIfNonEmpty(headers, "DNT", "0")
+	}
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    data,
+		Headers: headers,
+	}}, errs
+}
+
+func (a *ConnectAdAdapter) MakeBids(bidReq *openrtb.BidRequest, unused *adapters.RequestData, httpRes *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if httpRes.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if httpRes.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Invalid Status Returned: %d. Run with request.debug = 1 for more info", httpRes.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+
+	if err := json.Unmarshal(httpRes.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unable to unpackage bid response. Error: %s", err.Error()),
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &sb.Bid[i],
+				BidType: "banner",
+			})
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func preprocess(request *openrtb.BidRequest) []error {
+	impsCount := len(request.Imp)
+	errors := make([]error, 0, impsCount)
+	resImps := make([]openrtb.Imp, 0, impsCount)
+	secure := int8(0)
+
+	if request.Site != nil && request.Site.Page != "" {
+		pageURL, err := url.Parse(request.Site.Page)
+		if err == nil && pageURL.Scheme == "https" {
+			secure = int8(1)
+		}
+	}
+
+	for _, imp := range request.Imp {
+		cadExt, err := unpackImpExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		addImpInfo(&imp, &secure, cadExt)
+
+		if err := buildImpBanner(&imp); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		resImps = append(resImps, imp)
+	}
+
+	request.Imp = resImps
+
+	return errors
+}
+
+func addImpInfo(imp *openrtb.Imp, secure *int8, cadExt *openrtb_ext.ExtImpConnectAd) {
+	imp.TagID = strconv.Itoa(cadExt.SiteId)
+	imp.Secure = secure
+
+	if cadExt.Bidfloor != 0 {
+		imp.BidFloor = cadExt.Bidfloor
+		imp.BidFloorCur = "USD"
+	}
+
+	return
+}
+
+func addHeaderIfNonEmpty(headers http.Header, headerName string, headerValue string) {
+	if len(headerValue) > 0 {
+		headers.Add(headerName, headerValue)
+	}
+}
+
+func unpackImpExt(imp *openrtb.Imp) (*openrtb_ext.ExtImpConnectAd, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
```

Comment:
Consider including the impression index as you do with the other unmarshal failure branches.

Reply:
The code didn't change in this area. Do you feel that including the imp id here is not useful?